File: le-backend/run_e2e_tests.py
#!/usr/bin/env python3
"""
End-to-End Test Runner for System Stability Improvements

This script runs comprehensive end-to-end tests to validate all system
functionality and performance requirements.
"""

import asyncio
import sys
import os
import subprocess
import time
import json
from typing import Dict, List, Any, Optional
from pathlib import Path


class TestRunner:
    """Comprehensive test runner for end-to-end validation"""
    
    def __init__(self):
        self.results = {
            "start_time": time.time(),
            "test_suites": {},
            "summary": {
                "total_tests": 0,
                "passed": 0,
                "failed": 0,
                "skipped": 0,
                "errors": []
            }
        }
    
    def run_all_tests(self, include_slow: bool = False, include_stress: bool = False) -> bool:
        """Run all end-to-end test suites"""
        
        print("🚀 Starting End-to-End Test Suite for System Stability Improvements")
        print("=" * 80)
        
        # Test suites to run
        test_suites = [
            {
                "name": "Unit Tests",
                "command": ["python3", "-m", "pytest", "tests/", "-v", "--tb=short", "-m", "not slow"],
                "description": "Core unit tests for all components"
            },
            {
                "name": "Integration Tests", 
                "command": ["python3", "-m", "pytest", "tests/test_folder_integration.py", "tests/test_data_sync_integration.py", "-v"],
                "description": "Integration tests for component interactions"
            },
            {
                "name": "End-to-End Workflows",
                "command": ["python3", "-m", "pytest", "tests/test_end_to_end_workflows.py", "-v", "--tb=short"],
                "description": "Complete user workflow validation"
            },
            {
                "name": "Regression Tests",
                "command": ["python3", "-m", "pytest", "tests/test_regression_critical_paths.py", "-v", "--tb=short"],
                "description": "Critical path regression validation"
            }
        ]
        
        # Add performance tests if requested
        if include_slow:
            test_suites.append({
                "name": "Performance Tests",
                "command": ["python3", "-m", "pytest", "tests/test_performance_stress.py", "-v", "-m", "slow"],
                "description": "Performance and load testing"
            })
        
        # Add stress tests if requested
        if include_stress:
            test_suites.append({
                "name": "Stress Tests",
                "command": ["python3", "-m", "pytest", "tests/test_performance_stress.py::TestSystemResourceUsage", "-v"],
                "description": "System stress and resource usage testing"
            })
        
        all_passed = True
        
        for suite in test_suites:
            print(f"\n📋 Running {suite['name']}")
            print(f"   {suite['description']}")
            print("-" * 60)
            
            success = self._run_test_suite(suite)
            if not success:
                all_passed = False
                if not include_slow:  # Stop on failure for quick tests
                    break
        
        # Generate final report
        self._generate_report()
        
        return all_passed
    
    def _run_test_suite(self, suite: Dict[str, Any]) -> bool:
        """Run a single test suite"""
        
        start_time = time.time()
        
        try:
            # Run the test command
            result = subprocess.run(
                suite["command"],
                capture_output=True,
                text=True,
                timeout=600  # 10 minute timeout per suite
            )
            
            end_time = time.time()
            duration = end_time - start_time
            
            # Parse results
            success = result.returncode == 0
            
            suite_result = {
                "success": success,
                "duration": duration,
                "stdout": result.stdout,
                "stderr": result.stderr,
                "return_code": result.returncode
            }
            
            self.results["test_suites"][suite["name"]] = suite_result
            
            # Print summary
            if success:
                print(f"✅ {suite['name']} PASSED ({duration:.2f}s)")
                self._extract_test_counts(result.stdout, passed=True)
            else:
                print(f"❌ {suite['name']} FAILED ({duration:.2f}s)")
                print(f"   Return code: {result.returncode}")
                if result.stderr:
                    print(f"   Error: {result.stderr[:200]}...")
                self._extract_test_counts(result.stdout, passed=False)
                self.results["summary"]["errors"].append({
                    "suite": suite["name"],
                    "error": result.stderr[:500]
                })
            
            return success
            
        except subprocess.TimeoutExpired:
            print(f"⏰ {suite['name']} TIMEOUT (>10 minutes)")
            self.results["test_suites"][suite["name"]] = {
                "success": False,
                "duration": 600,
                "error": "Test suite timed out"
            }
            self.results["summary"]["errors"].append({
                "suite": suite["name"],
                "error": "Test suite timed out after 10 minutes"
            })
            return False
            
        except Exception as e:
            print(f"💥 {suite['name']} ERROR: {e}")
            self.results["test_suites"][suite["name"]] = {
                "success": False,
                "duration": time.time() - start_time,
                "error": str(e)
            }
            self.results["summary"]["errors"].append({
                "suite": suite["name"],
                "error": str(e)
            })
            return False
    
    def _extract_test_counts(self, output: str, passed: bool) -> None:
        """Extract test counts from pytest output"""
        
        # Look for pytest summary line
        lines = output.split('\n')
        for line in lines:
            if 'passed' in line or 'failed' in line or 'error' in line:
                # Try to extract numbers
                import re
                numbers = re.findall(r'(\d+) (\w+)', line)
                for count, status in numbers:
                    count = int(count)
                    if status in ['passed', 'PASSED']:
                        self.results["summary"]["passed"] += count
                    elif status in ['failed', 'FAILED', 'error', 'ERROR']:
                        self.results["summary"]["failed"] += count
                    elif status in ['skipped', 'SKIPPED']:
                        self.results["summary"]["skipped"] += count
                    
                    self.results["summary"]["total_tests"] += count
    
    def _generate_report(self) -> None:
        """Generate comprehensive test report"""
        
        end_time = time.time()
        total_duration = end_time - self.results["start_time"]
        
        print("\n" + "=" * 80)
        print("📊 END-TO-END TEST REPORT")
        print("=" * 80)
        
        # Summary
        summary = self.results["summary"]
        print(f"Total Duration: {total_duration:.2f} seconds")
        print(f"Total Tests: {summary['total_tests']}")
        print(f"✅ Passed: {summary['passed']}")
        print(f"❌ Failed: {summary['failed']}")
        print(f"⏭️  Skipped: {summary['skipped']}")
        
        # Success rate
        if summary['total_tests'] > 0:
            success_rate = (summary['passed'] / summary['total_tests']) * 100
            print(f"Success Rate: {success_rate:.1f}%")
        
        # Suite breakdown
        print("\n📋 Test Suite Results:")
        for suite_name, suite_result in self.results["test_suites"].items():
            status = "✅ PASS" if suite_result["success"] else "❌ FAIL"
            duration = suite_result.get("duration", 0)
            print(f"  {status} {suite_name} ({duration:.2f}s)")
        
        # Errors
        if summary["errors"]:
            print("\n🚨 Errors:")
            for error in summary["errors"]:
                print(f"  - {error['suite']}: {error['error'][:100]}...")
        
        # Save detailed report
        report_file = "test_report.json"
        with open(report_file, 'w') as f:
            json.dump(self.results, f, indent=2)
        print(f"\n📄 Detailed report saved to: {report_file}")
        
        # Performance summary
        self._print_performance_summary()
    
    def _print_performance_summary(self) -> None:
        """Print performance test summary if available"""
        
        perf_suite = self.results["test_suites"].get("Performance Tests")
        if not perf_suite:
            return
        
        print("\n⚡ Performance Summary:")
        
        if perf_suite["success"]:
            print("  ✅ All performance tests passed")
            print("  📈 System meets performance requirements")
        else:
            print("  ❌ Some performance tests failed")
            print("  ⚠️  System may not meet performance requirements")
        
        # Extract performance metrics from output if available
        if "stdout" in perf_suite:
            output = perf_suite["stdout"]
            
            # Look for performance metrics
            lines = output.split('\n')
            for line in lines:
                if 'uploads/sec' in line or 'MB/s' in line or 'ops/sec' in line:
                    print(f"  📊 {line.strip()}")


def main():
    """Main entry point"""
    
    import argparse
    
    parser = argparse.ArgumentParser(
        description="Run comprehensive end-to-end tests for system stability improvements"
    )
    parser.add_argument(
        "--include-slow", 
        action="store_true",
        help="Include slow performance tests"
    )
    parser.add_argument(
        "--include-stress",
        action="store_true", 
        help="Include stress tests (requires --include-slow)"
    )
    parser.add_argument(
        "--quick",
        action="store_true",
        help="Run only quick tests (unit + integration)"
    )
    
    args = parser.parse_args()
    
    # Validate arguments
    if args.include_stress and not args.include_slow:
        print("❌ Error: --include-stress requires --include-slow")
        sys.exit(1)
    
    # Change to backend directory
    backend_dir = Path(__file__).parent
    os.chdir(backend_dir)
    
    # Check if we're in a virtual environment
    if not os.environ.get('VIRTUAL_ENV') and not sys.prefix != sys.base_prefix:
        print("⚠️  Warning: Not running in a virtual environment")
        print("   Consider activating your virtual environment first")
    
    # Run tests
    runner = TestRunner()
    
    if args.quick:
        # Quick test mode - only unit and integration tests
        success = runner.run_all_tests(include_slow=False, include_stress=False)
    else:
        success = runner.run_all_tests(
            include_slow=args.include_slow,
            include_stress=args.include_stress
        )
    
    # Exit with appropriate code
    if success:
        print("\n🎉 All tests passed! System is stable and ready for deployment.")
        sys.exit(0)
    else:
        print("\n💥 Some tests failed. Please review the errors above.")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: lc-workflow-frontend/run_e2e_tests.js
#!/usr/bin/env node

/**
 * Frontend End-to-End Test Runner
 * 
 * Runs comprehensive frontend tests to validate user interface
 * functionality and integration with backend services.
 */

const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

class FrontendTestRunner {
  constructor() {
    this.results = {
      startTime: Date.now(),
      testSuites: {},
      summary: {
        totalTests: 0,
        passed: 0,
        failed: 0,
        skipped: 0,
        errors: []
      }
    };
  }

  async runAllTests(options = {}) {
    console.log('🚀 Starting Frontend End-to-End Test Suite');
    console.log('='.repeat(80));

    const testSuites = [
      {
        name: 'Unit Tests',
        command: 'npm',
        args: ['test', '--', '--watchAll=false', '--coverage=false', '--testPathPattern=(?!.*e2e).*\\.test\\.(ts|tsx|js|jsx)$'],
        description: 'Component and utility unit tests'
      },
      {
        name: 'Integration Tests',
        command: 'npm',
        args: ['test', '--', '--watchAll=false', '--testPathPattern=.*__tests__.*Integration.*\\.test\\.(ts|tsx)$'],
        description: 'Component integration tests'
      },
      {
        name: 'E2E Workflow Tests',
        command: 'npm',
        args: ['test', '--', '--watchAll=false', '--testPathPattern=.*e2e.*\\.test\\.(ts|tsx)$'],
        description: 'End-to-end user workflow tests'
      }
    ];

    // Add performance tests if requested
    if (options.includePerformance) {
      testSuites.push({
        name: 'Performance Tests',
        command: 'npm',
        args: ['test', '--', '--watchAll=false', '--testPathPattern=.*performance.*\\.test\\.(ts|tsx)$'],
        description: 'Frontend performance and load tests'
      });
    }

    let allPassed = true;

    for (const suite of testSuites) {
      console.log(`\n📋 Running ${suite.name}`);
      console.log(`   ${suite.description}`);
      console.log('-'.repeat(60));

      const success = await this.runTestSuite(suite);
      if (!success) {
        allPassed = false;
        if (!options.continueOnFailure) {
          break;
        }
      }
    }

    // Generate report
    this.generateReport();

    return allPassed;
  }

  async runTestSuite(suite) {
    const startTime = Date.now();

    try {
      const result = await this.executeCommand(suite.command, suite.args, {
        timeout: 300000 // 5 minute timeout
      });

      const endTime = Date.now();
      const duration = (endTime - startTime) / 1000;

      const success = result.code === 0;

      const suiteResult = {
        success,
        duration,
        stdout: result.stdout,
        stderr: result.stderr,
        returnCode: result.code
      };

      this.results.testSuites[suite.name] = suiteResult;

      if (success) {
        console.log(`✅ ${suite.name} PASSED (${duration.toFixed(2)}s)`);
        this.extractTestCounts(result.stdout, true);
      } else {
        console.log(`❌ ${suite.name} FAILED (${duration.toFixed(2)}s)`);
        console.log(`   Return code: ${result.code}`);
        if (result.stderr) {
          console.log(`   Error: ${result.stderr.substring(0, 200)}...`);
        }
        this.extractTestCounts(result.stdout, false);
        this.results.summary.errors.push({
          suite: suite.name,
          error: result.stderr.substring(0, 500)
        });
      }

      return success;

    } catch (error) {
      console.log(`💥 ${suite.name} ERROR: ${error.message}`);
      
      const duration = (Date.now() - startTime) / 1000;
      
      this.results.testSuites[suite.name] = {
        success: false,
        duration,
        error: error.message
      };

      this.results.summary.errors.push({
        suite: suite.name,
        error: error.message
      });

      return false;
    }
  }

  executeCommand(command, args, options = {}) {
    return new Promise((resolve, reject) => {
      const child = spawn(command, args, {
        stdio: ['pipe', 'pipe', 'pipe'],
        shell: true
      });

      let stdout = '';
      let stderr = '';

      child.stdout.on('data', (data) => {
        const output = data.toString();
        stdout += output;
        // Show real-time output for better feedback
        if (!options.quiet) {
          process.stdout.write(output);
        }
      });

      child.stderr.on('data', (data) => {
        const output = data.toString();
        stderr += output;
        if (!options.quiet) {
          process.stderr.write(output);
        }
      });

      child.on('close', (code) => {
        resolve({
          code,
          stdout,
          stderr
        });
      });

      child.on('error', (error) => {
        reject(error);
      });

      // Handle timeout
      if (options.timeout) {
        setTimeout(() => {
          child.kill('SIGTERM');
          reject(new Error(`Command timed out after ${options.timeout}ms`));
        }, options.timeout);
      }
    });
  }

  extractTestCounts(output, passed) {
    // Extract Jest test results
    const lines = output.split('\n');
    
    for (const line of lines) {
      // Look for Jest summary lines
      const testMatch = line.match(/Tests:\s+(\d+)\s+failed,\s+(\d+)\s+passed,\s+(\d+)\s+total/);
      if (testMatch) {
        const [, failed, passedCount, total] = testMatch;
        this.results.summary.failed += parseInt(failed);
        this.results.summary.passed += parseInt(passedCount);
        this.results.summary.totalTests += parseInt(total);
        return;
      }

      // Alternative format
      const altMatch = line.match(/(\d+)\s+passing/);
      if (altMatch) {
        const [, passedCount] = altMatch;
        this.results.summary.passed += parseInt(passedCount);
        this.results.summary.totalTests += parseInt(passedCount);
      }

      const failMatch = line.match(/(\d+)\s+failing/);
      if (failMatch) {
        const [, failedCount] = failMatch;
        this.results.summary.failed += parseInt(failedCount);
        this.results.summary.totalTests += parseInt(failedCount);
      }
    }
  }

  generateReport() {
    const endTime = Date.now();
    const totalDuration = (endTime - this.results.startTime) / 1000;

    console.log('\n' + '='.repeat(80));
    console.log('📊 FRONTEND TEST REPORT');
    console.log('='.repeat(80));

    // Summary
    const summary = this.results.summary;
    console.log(`Total Duration: ${totalDuration.toFixed(2)} seconds`);
    console.log(`Total Tests: ${summary.totalTests}`);
    console.log(`✅ Passed: ${summary.passed}`);
    console.log(`❌ Failed: ${summary.failed}`);
    console.log(`⏭️  Skipped: ${summary.skipped}`);

    // Success rate
    if (summary.totalTests > 0) {
      const successRate = (summary.passed / summary.totalTests) * 100;
      console.log(`Success Rate: ${successRate.toFixed(1)}%`);
    }

    // Suite breakdown
    console.log('\n📋 Test Suite Results:');
    for (const [suiteName, suiteResult] of Object.entries(this.results.testSuites)) {
      const status = suiteResult.success ? '✅ PASS' : '❌ FAIL';
      const duration = suiteResult.duration || 0;
      console.log(`  ${status} ${suiteName} (${duration.toFixed(2)}s)`);
    }

    // Errors
    if (summary.errors.length > 0) {
      console.log('\n🚨 Errors:');
      for (const error of summary.errors) {
        console.log(`  - ${error.suite}: ${error.error.substring(0, 100)}...`);
      }
    }

    // Save detailed report
    const reportFile = 'frontend-test-report.json';
    fs.writeFileSync(reportFile, JSON.stringify(this.results, null, 2));
    console.log(`\n📄 Detailed report saved to: ${reportFile}`);

    // Coverage information
    this.printCoverageInfo();
  }

  printCoverageInfo() {
    console.log('\n📈 Test Coverage:');
    
    // Check if coverage directory exists
    const coverageDir = path.join(process.cwd(), 'coverage');
    if (fs.existsSync(coverageDir)) {
      console.log('  ✅ Coverage reports generated in ./coverage/');
      console.log('  🌐 Open ./coverage/lcov-report/index.html to view detailed coverage');
    } else {
      console.log('  ℹ️  Run with --coverage to generate coverage reports');
    }
  }
}

// Test configuration validation
function validateTestEnvironment() {
  const requiredFiles = [
    'package.json',
    'jest.config.js',
    'jest.setup.js'
  ];

  const missingFiles = requiredFiles.filter(file => !fs.existsSync(file));
  
  if (missingFiles.length > 0) {
    console.error('❌ Missing required test configuration files:');
    missingFiles.forEach(file => console.error(`   - ${file}`));
    process.exit(1);
  }

  // Check if node_modules exists
  if (!fs.existsSync('node_modules')) {
    console.error('❌ node_modules not found. Run "npm install" first.');
    process.exit(1);
  }

  console.log('✅ Test environment validation passed');
}

// Main execution
async function main() {
  const args = process.argv.slice(2);
  
  const options = {
    includePerformance: args.includes('--include-performance'),
    continueOnFailure: args.includes('--continue-on-failure'),
    quick: args.includes('--quick')
  };

  // Show help
  if (args.includes('--help') || args.includes('-h')) {
    console.log(`
Frontend End-to-End Test Runner

Usage: node run_e2e_tests.js [options]

Options:
  --include-performance    Include performance tests
  --continue-on-failure    Continue running tests even if some fail
  --quick                  Run only quick tests (unit + integration)
  --help, -h              Show this help message

Examples:
  node run_e2e_tests.js                    # Run standard test suite
  node run_e2e_tests.js --quick            # Quick test run
  node run_e2e_tests.js --include-performance  # Include performance tests
`);
    process.exit(0);
  }

  console.log('🔍 Validating test environment...');
  validateTestEnvironment();

  const runner = new FrontendTestRunner();
  
  try {
    const success = await runner.runAllTests(options);
    
    if (success) {
      console.log('\n🎉 All frontend tests passed! UI is stable and ready.');
      process.exit(0);
    } else {
      console.log('\n💥 Some frontend tests failed. Please review the errors above.');
      process.exit(1);
    }
  } catch (error) {
    console.error('\n💥 Test runner encountered an error:', error.message);
    process.exit(1);
  }
}

// Handle unhandled promise rejections
process.on('unhandledRejection', (reason, promise) => {
  console.error('Unhandled Rejection at:', promise, 'reason:', reason);
  process.exit(1);
});

// Run if called directly
if (require.main === module) {
  main();
}

module.exports = { FrontendTestRunner };